float64 giro
float64 toggle
